### PR TITLE
Populate heap when periodic tasks are changed - fixed PEP8

### PR DESCRIPTION
**Pull request by gitaarik**
I improved @wzywno's pullrequest:
https://github.com/celery/celery/pull/3791

I fixed the PEP8 issues, hope it can be merged soon.

I also squashed the commits together in one, so there's a clean history.

### CODE REVIEW

**Comment by AJRenold** [comment]
Question on this solution - it seems a bit inefficient to check if the schedule has changed every time `tick` is called. Why not add a method which is called whenever the schedule changes? I've only worked with the [django-celery-beat ](https://github.com/celery/django-celery-beat) beat command, but that is the change which I made to resolve this issue - https://github.com/celery/django-celery-beat/blob/master/django_celery_beat/schedulers.py#L289 - setting Schedule._heap to None was not the cleanest way to do this because _heap is 'private' to the parent class, but it was the only way I saw to invalidate the heap when the schedule was updated

**Comment by brianmay** [comment]
This doesn't seem to resolve the problem for me.

**Comment by brianmay** [comment]
My initial test was flawed, this PR does solve celery/django-celery-beat#7.

**Comment by brianmay** [comment]
@AJRenold I have tried the master branch, which includes your change - I checked to make sure, and it doesn't resolve the issue.

**Comment by AJRenold** [comment]
Hey, interesting! Do you think we can write a failing test on django-celery-beat? What exactly are you changing in your PeriodicTask object, is it one of IntervalSchedule, CrontabSchedule or SolarSchedule?

**Comment by brianmay** [comment]
I think I was changing the PeriodicTask by pointing it to a different IntervalSchedule.

Yes, new test cases is always a good idea.

File: celery/beat.py [hunk 1]
```diff
@@ -257,7 +258,9 @@ def tick(self, event_t=event_t, min=min, heappop=heapq.heappop,
         adjust = self.adjust
         max_interval = self.max_interval
 
-        if self._heap is None:
+        if (self._heap is None or
+                not self.schedules_equal(self.old_schedulers, self.schedule)):
```
Comment (georgepsarakis) [inline celery/beat.py]: The [schedule](https://github.com/celery/celery/blob/master/celery/schedules.py#L171) class has defined the `_eq__` method, so I think you can compare directly the dictionaries `self.old_schedulers` and `self.schedule`, you should not need an additional function to achieve this.

File: celery/beat.py [hunk 2]
```diff
@@ -257,7 +258,9 @@ def tick(self, event_t=event_t, min=min, heappop=heapq.heappop,
         adjust = self.adjust
         max_interval = self.max_interval
 
-        if self._heap is None:
+        if (self._heap is None or
```
Comment (georgepsarakis) [inline celery/beat.py]: Could you please add a test case verifying the flow for this logical branch?

**Comment by gitaarik** [comment]
To be honest I don't know about the logic in these changes, I just made it PEP8. @wzywno, do you care to take a closer look?

**Comment by brianmay** [comment]
Just noticed this change only works if you change the periodic task. If you change the interval object, the new value is not reloaded. Will retest again against the master version, make sure this wasn't confusing me.

I also have a branch based on this one that uses ``!=`` instead of ``schedules_equal``, which appears to work fine.

**Comment by brianmay** [comment]
Ok, reconfirmed that master of django-celery-beat by itself does not fix this.

File: celery/beat.py (hunk 1 above)
Comment (brianmay) [inline celery/beat.py]: See https://github.com/celery/django-celery-beat/issues/7#issuecomment-267451710 for why this apparently ugly code is required.

**Comment by brianmay** [comment]
Just deleted a comment I just made because it was wrong/misleading.

I created a pull request to address the feedback here, but found that it broke. As documented here https://github.com/celery/django-celery-beat/issues/7#issuecomment-267451710

With the simplified code, changes the the interval table are not detected.

File: celery/beat.py [hunk 3]
```diff
@@ -281,6 +284,18 @@ def tick(self, event_t=event_t, min=min, heappop=heapq.heappop,
                 return min(verify[0], max_interval)
         return min(adjust(next_time_to_run) or max_interval, max_interval)
 
+    def schedules_equal(self, a, b):
+        if a.keys() != b.keys():
```
Comment (brianmay) [inline celery/beat.py]: Yes, that does appear wrong. Will try to understand how this code works.

Comment (brianmay) [inline celery/beat.py]: ``set(a.keys()) != set(b.keys())`` works for me.

File: celery/beat.py [hunk 4]
```diff
@@ -281,6 +284,18 @@ def tick(self, event_t=event_t, min=min, heappop=heapq.heappop,
...
+            return False
+        for name, model in a.items():
+            b_model = b.get(name)
+            if not b_model:
+                return False
+            if (hasattr(model.schedule, '__repr__') and
+                    model.schedule.__repr__() != b_model.schedule.__repr__()):
```
Comment (brianmay) [inline celery/beat.py]: I believe this if comparison can be simplified to:

```
if model.schedule != b_model.schedule:
```

**Comment by brianmay** [comment]
celery/celery#3958 works for me.

File: celery/beat.py (hunk 4 above)
Comment (brianmay) [inline celery/beat.py]: `model` and `b_model` are of type `django_celery_beat.schedulers.ModelEntry`.

So this code is django_celery_beat specific. Not convinced this is appropriate code for celery :-(

File: celery/beat.py (hunk 1 above)
Comment (brianmay) [inline celery/beat.py]: These `self.old_schedulers` and `self.schedule` are of type `dict`, so the `==` or `!=` is only a shallow comparison.

File: celery/beat.py (hunk 4 above)
Comment (brianmay) [inline celery/beat.py]: Oh, wait, my objection to my own pull request might be unfounded.

`django_celery_beat.schedulers.ModelEntry` is actually inherited from `celery.beat ScheduleEntry` - so I am guessing is actually a standard celery object with some extra bits added.

**Comment by subha-py** [comment]
Can we have a working example and celery beat/ celery link? I am not able to run scheduled tasks with django 1.11

**Comment by brianmay** [comment]
@subha-py Django 1.11 support works for me... Maybe open another bug report with full details?

**Comment by brianmay** [comment]
I think this pull request can now be closed, it was merged in #3958.

**Pull request closed without merging**